### DISCUSSION TITLE
Storing JSON with arbitrary keys

### DISCUSSION THREAD

**Discussion by dwoznicki** [ClickHouse/ClickHouse#48916] [Q&A] [answered]
Hello, I'm using ClickHouse as a data store for metrics data, and looking for the best way to store arbitrary JSON objects that can be associated with each data point.

Data is modeled such that each data point belongs to a metric. A data point has a time, numeric value, and metadata (represented as collection of key-value pairs, basically a JSON object). For example, a metric "Metric A" might have a row like this.

```
(2022-01-01 00:00:00, 99.0, {"host": "my_hostname", "version": 11})
```

I'm having trouble figuring out what to do with the metadata object. I'll need to filter on it sometimes (e.g. `SELECT * FROM table WHERE metadata.host = 'my_hostname'`), and I'd like to preserve the value type so I can query on that as well (e.g. `SELECT * FROM table WHERE metadata.version > 10`). I'll also need to support a somewhat arbitrary set of keys. I think it's reasonable to impose a cardinality limit on keys per metric (e.g. "Metric A" metadata can have no more than 500 distinct keys), but I don't want to impose such a limit on all metrics in the database (e.g. metadata for all metrics can have no more than 500 distinct keys).

Here's what I've considered so far.

- **Store metadata as a JSON string.** This is super flexible, and ClickHouse has excellent support for working with JSON strings. But it sounds like this can be slow as ClickHouse needs to do a full table scan for JSON string filtering.
- **Store metadata as the new JSON type.** I'm a bit hesitant since the JSON type is marked as experimental still. Also, it looks like the JSON type is a dynamically updated tuple under the hood, which means that whenever a new key is added, every metadata tuple in the table is updated with a default value for the key, even if the original metadata object did not contain that key. I'm also worried about hitting a tuple max length limit.
- **Store metadata as a tuple and dynamically update the tuple when a new key is found.** Basically, I could mock the behavior of the JSON type myself, but allow for `Nullable` tuple values to get around the default value problem. Again, I'm worried about tuple length.
- **Store metadata as a map(String, String).** This seems good, but I lose the ability to query based on original type. I might be willing to make that trade-off, would prefer to find another way.
- **Split metric data into multiple, dynamically generated tables, one per metric, and store metadata as a dynamically updated tuple.** This satisfies all my requirements mentioned above, but I'm concerned that I'm going to end up with tons of small inserts to individual tables, [which is a ClickHouse anti-pattern](https://clickhouse.com/blog/common-getting-started-issues-with-clickhouse). I also have a slight preference for not using dynamically generated tables for maintainability.

I'd appreciate any advice/insight that anyone can provide.

**Comment by gingerwizard** [comment, marked as answer]
Hi @dwoznicki 

Long term the JSON type is the solution here. However, its currently experiemental as you point out and not something i would rely on right now.

I think you summarize the tradeoffs quite well. Personally, i would lean towards using the `Map(LowCardinality(String),String)` type - this is what the OTEL collector does. I would combine this with [materialized columns](https://clickhouse.com/docs/en/sql-reference/statements/create/table#materialized) for keys i query frequently. This avoids the need to do type conversions in your queries but will also provide significant speedup - mainly as the map column is completely loaded to access just one key. This can impact query performance appreciably.

The OTEL collector also uses bloom filters for the Map keys - [more on otel here](https://clickhouse.com/blog/storing-traces-and-spans-open-telemetry-in-clickhouse). I've no first hand experience of their effectiveness but this largely aligns with [best practices](https://clickhouse.com/docs/en/optimize/skipping-indexes) - especially if your key occurences align with your primary keys.
